text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+
+union MyUnion<T: Copy> {
+    uninit: (),
+    value: (T, !),
+}
+impl<T: Copy> MyUnion<T> {
+    unsafe fn assume_init(self) -> ! {
+        match self { MyUnion { value: (_, !) } }
+    }
+}
+```
+
+<!--
+
+```rust
+enum E { A(i32), B(i32, !) }
+let result: *const E = some_function();
+unsafe {
+    let x = match *result {
+        A(x) => x,
+        // An arm is required (https://github.com/rust-lang/unsafe-code-guidelines/issues/443) but `!` asserts validity of the `!` data so the arm doesn't need a body.
+        B(_, !),
+    };
+    // Alternatively:
+    let (A(x) | B(_, !)) = *result;
+}
+```
+
+```rust
+enum E { A(i32), B(i32, !) }
+let val: *mut E = Box::leak(Box::new(E::A(42)));
+// It may be possible to partially initialize this to the `B` variant
+// in the future.
+unsafe { set_discriminant_to_b(val) };
+unsafe {
+    match *val {
+        A(x) => { ... }
+        // This branch would then be reached without UB.
+        B(..) => println!("Reachable!"),
+    }
+}
+
+// It may be possible to construct a `&!` pointing to
+// uninitialized data (unsafe, but valid).
+let never_ref: &! = ...;
+let result: Result<u8, &!> = Err(never_ref);
+match result {
+    Ok(x) => { ... }
+    // This branch would then be reached without UB.
+    Err(_) => println!("Reachable!"),
+}
+```
+
+```rust
+enum Void {}
+let result: &Result<T, Void> = ...;
+let x = match result {
+    Ok(x) => x,
+    Err(!), // No need for an arm body
+};
+// Or even
+let (Ok(x) | Err(!)) = result;
+```
+
+-->
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Patterns can be used on a partially uninitialized place; the basic rule is that a pattern only accesses data that is directly mentioned in the pattern: an `Ok(..)` pattern requires a discriminant read, a binding requires a full read, a wildcard `_` accesses nothing etc. 
+
+For uninhabited types (types with no valid values), normal patterns cannot be used to indicate an access since there is no data to be accessed. For this purpose, you can use the special `!` pattern.
+
+The `!` pattern is allowed on any uninhabited type (such as `enum Void {}` or `(Result<!, !>, u32)`) and does an access to the underlying data. Since there can be no such data (on pain of UB), the corresponding arm is known to be unreachable and does not need a body:
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We add `!` to the syntax of patterns. A `!` pattern is accepted for any type that is visibly uninhabited (i.e. uninhabited taking into account private fields and `#[non_exhaustive]` annotations) from the item containing the match expression.
```

Comment:
There are many possible definitions of "uninhabited" so it'd be good to spell this out. E.g., is `&!` uninhabited?

What I would expect is that `!` is allowed on all types whose *safety* invariant cannot be satisfied (it is an explicit operation after all), so we can treat `&!` as uninhabited.

Reply:
Oh yeah, I had not considered the safety version of uninhabitedness. This does feel like the appropriate notion to use, I'll change that

Reply:
More generally, would the `!` pattern be able to multiple uninhabited types? Stretching this to extremes, would the following compile?
```rust
enum Empty {}
enum MultiNever<T> {
   A(T),
   B(Empty),
   C(Empty),
}
fn abc(value: MultiNever<&MultiNever<u32>>) -> u32 {
   match value {
       MultiNever::A(MultiNever::A(value)) => *value,
       !
   }
}
```

Reply:
Just revisiting my musing in the last comment, I guess really the motivation in the RFC is actually - more verbosely - that "accesses which don't violate any safety invariant and could cause UB should be explicitly visible in the pattern".

And so Ralf's justification is that, references have a safety invariant to point at valid data, therefore it should be okay to use them. But e.g. pointers or union fields do not have such a safety invariant, so would not satisfy such a recursive safety-based definition of uninhabitedness.

Reply:
> Just to clarify, if we use this wider definition of "uninhabited" for when the `!` pattern can apply, does this mean the following would be acceptable?

No, this is only about whether `!` would apply to e.g. `(u32, &!)`. I chose not to allow what you suggest because 1. it fails at being explicit which is the point of never patterns 2. it's a lot more annoying to implement and specify.

> By contrast, e.g. pointers or union fields do not have such a safety invariant, so would not satisfy such a recursive safety-based definition of uninhabitedness, and would need to be handled explicitly.

Yep

> It appears to me that (u32, !) has a safety variant that it can't be satisfied, and so would come under the suggested definition of a never pattern.

Under both definitions (the "safety" one and the "validity" one, this is considered inhabited, and thus a `!` pattern would be allowed on that type.

> Are we happy that a blanket ! could assert such types are invalid, without requiring a user to spell it out explicitly as e.g. a (v, !) pattern?

This is what I propose in the RFC, given the understanding that a `!` pattern asserts "there can be no safe value here".

Reply:
> I still think we should allow this. `!` is quite explicit, so IMO if the safety invariant of a type is unsatisfiable, then we should allow using the `!` pattern on it.

Yeah that's absolutely my plan :) I'll update the RFC when I get a moment.

Reply:
I seem to have not been clear in my replies, allow me to clarify. My intent is still: "A `!` pattern is accepted for any type that is visibly uninhabited".

What I meant by "uninhabited" when I wrote the RFC was: "a type that has no values that satisfy it's _validity_ invariant (taking into account private fields etc)". Per Ralf's suggestion, I now prefer the following definition: "a type that has no values that satisfy it's _safety_ invariant (taking into account private fields etc)".

Both these definitions are recursive. The only difference is how they handle `&T`: the validity version ([conservatively?](https://github.com/rust-lang/unsafe-code-guidelines/issues/413)) considers `&T` always inhabited; the safety version considers `&T` inhabited exactly when `T` is inhabited.

Reply:
When I said "I chose not to allow what you suggest", I meant specifically things like:
```rust
match foo {
    None => ...,
    !
}
```
I prefer that the user write `Some(!)` explicitly, because this is more explicit and easier to specify/implement.

Reply:
Thanks for the clarification! Apologies - I think my confusion stemmed from my incorrectly conflating (A) whether `!` can wildcard match multiple branches/types with (B) the notion of recursive inhabitedness.

So just to clarify, I believe the that the first 3 compile, but the fourth one does not:
```rust
enum Empty {}
enum MyEnum<T, T2> {
    A(T),
    B(T2),
    C(Empty),
}

fn fn1(value: MyEnum<&Empty, Empty>) -> u32 {
    match value { ! }
}
fn fn2(value: MyEnum<MyEnum<Empty, &(u64, Empty)>, Empty>) -> u32 {
    match value { ! }
}
fn fn3(value: MyEnum<&MyEnum<u32, Empty>, &MyEnum<Empty, Empty>>) -> u32 {
    match value {
        MultiNever::A(a) => match a {
            MultiNever::A(out) => *out,
            MultiNever::B(!),
            MultiNever::C(!),
        }
        MultiNever::B(!),
        MultiNever::C(!),
    }
}
// DOES NOT COMPILE
fn fn4(value: MyEnum<u32, Empty>) -> u32 {
    match value {
        MyEnum::A(a) => a,
        !
    }
}

```

Reply:
> I prefer that the user write Some(!) explicitly because I find it more explicit and easier to specify/implement. This is open for discussion of course, that's the aim of the RFC.

Good point! Well, in the spirit of discussion, let me propose for debate the `!` uninhabitable binding definition of the never pattern. Such a pattern would:
> Apply against all unmatched branches (similarly to `_` or a variable binding `a`), and, both:
> * Assert (on pain of a compile error) that all such types are safety-uninhabited
> * Assert (on pain of UB) that all such branches are not reachable (i.e. all such types are validity-uninhabited in this position)

If this is well defined, it appears to me that it's a strictly more powerful extension to this proposal - in the sense that it is allowed in a superset of places.

I was also wondering it it were too powerful, but I think on balance I don't think it is. It also has benefits that it can partly solve the macro ergonomics issue without auto-never, by blanket-adding a `!` to all generated matches, and adding something like an `#[allow(unneeded_never_pattern)]`.

Reply:
> So just to clarify, I believe that under this RFC, the first 3 compile, but the fourth one does not:

correct!

> If this is well defined (and, to be fair, that's a big if), it appears to me that it's a strictly more powerful extension to this proposal - in the sense that it is allowed in a superset of places.

I believe this is well-defined, and agree that this would be allowed in strictly more places than the RFC currently proposes.

> It has benefits that it can mostly solve the ergonomics issues in https://github.com/rust-lang/rust/issues/131452 without auto-never

If that's the main motivation I'd much rather have auto-never, because I'd much rather avoid safe users needing to add `!` patterns in some matches and not others for hard-to-figure-out reasons (namely the presence of `&` on the path to an empty type).

I tried to separate this RFC from auto-never discussions but maybe they're too entangled for that :thinking:.

Reply:
> If that's the main motivation I'd much rather have auto-never, because I'd much rather avoid safe users needing to add ! patterns in some matches and not others for hard-to-figure-out reasons (namely the presence of & on the path to an empty type).

I agree that we need an auto-never equivalent... As you've said in the comment below, maybe we need to explore that first.

But I think a more flexible never pattern could theoretically be beneficial to people writing unsafe too. For example, in these quite hypothetical scenarios:
* Maybe in their current circumstance they know that all branches that are unsafe are invalid, and want to avoid writing out lots of distinct branches. (e.g. they're using something like an unsafe [Trees That Grow](https://github.com/rust-lang/rust/issues/131452#issuecomment-2468750921) pattern, if such a thing would even make sense... in which case, they could use `!` as a wildcard, [like Ralf spelled out here](https://github.com/rust-lang/rust/issues/131452#issuecomment-2468883553). Although I'm not sure if that was intentional on Ralf's part, or accidental 🤷)
* Or they know that some sub-enum should always be never, and want to assert that at compile time.

Whilst I'm naturally drawn to neat generalisations, I accept that the use case argument isn't super strong.

That said, (ignoring the implementation difficulty argument for now) I feel a good argument against a more general / flexible pattern should be motivated in why a restriction makes sense. For example:
* Do we feel that allowing wildcard `!` is a foot gun for a user to write the wrong thing? That they put `!` as an easy fix because they don't understand the implications about it matching more variants than expected, or future new variants? (I'm not sure to what extent unsafe developers would benefit from such hand-holding)
* Do we feel that allowing wildcard ! might be "less easy to understand" or less readable at code review time? Does a lone `!` look weird or eat up too much strangeness budget?

Reply:
Good point, I didn't realize that we'd need more than just this RFC for that case.

My immediate gut feeling is that the conservative choice here is to accept less code. It's east to accept more code in the future, it's impossible to take back accepting wildcard `!` if we add it now. So I'd be inclined to first require the `!` pattern to be used in a position where the type is actually uninhabited, and possibly change the "automatic never pattern" rules for `match`es on entirely safe places, and *then* revisit whether we really need wildcard `!`.

Reply:
I think that's a good argument. Start conservative, without precluding expanding in future if it's still useful :+1: